internal/transports/p2p/server.go

```diff
@@ -29,99 +43,271 @@ func NewServer(
+		return errors.New("cannot connect to any peers from seed")
 	}
 
-	inbound := false
-	conn, err := net.Dial(firstPeerAddr.Network(), firstPeerAddr.String())
+	s.log.Info().Msgf("connected to %d peers", peersCounter)
+	go s.observeOutboundPeers()
+	return nil
+}
+
+func (s *server) listenInboundPeers() error {
+	ourAddr := net.JoinHostPort("", fmt.Sprintf("%d", s.chainParams.DefaultPort))
+	listener, err := net.Listen("tcp", ourAddr)
 	if err != nil {
-		s.log.Error().Msgf("error connecting to peer %s, reason: %v", firstPeerAddr.String(), err)
+		s.log.Error().Msgf("error creating listener, reason: %v", err)
 		return err
 	}
 
-	return s.connectPeer(conn, inbound)
+	s.listener = listener
+	go s.observeInboundPeers()
+	return nil
 }
 
-func (s *server) listenAndConnect() error {
-	s.log.Info().Msgf("listening for inbound connections on port %s", s.chainParams.DefaultPort)
+func (s *server) connectToAddr(addr net.IP, port uint16) error {
+	netAddr := &net.TCPAddr{
+		IP:   addr,
+		Port: int(port),
+	}
 
-	ourAddr := net.JoinHostPort("", s.chainParams.DefaultPort)
-	listener, err := net.Listen("tcp", ourAddr)
+	conn, err := net.Dial(netAddr.Network(), netAddr.String())
 	if err != nil {
-		s.log.Error().Msgf("error creating listener, reason: %v", err)
+		s.log.Error().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("error connecting with peer, reason: %v", err)
+
+		s.log.Info().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("peer banned, reason: %v", err)
+
+		s.addresses.BanAddr(&wire.NetAddress{IP: netAddr.IP, Port: uint16(netAddr.Port)})
 		return err
 	}
 
-	inbound := true
-	conn, err := listener.Accept()
+	peer, err := s.connectPeer(conn, false)
 	if err != nil {
-		s.log.Error().Msgf("error accepting connection, reason: %v", err)
+		s.log.Info().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("peer banned, reason: %v", err)
+
+		s.addresses.BanAddr(&wire.NetAddress{IP: netAddr.IP, Port: uint16(netAddr.Port)})
 		return err
 	}
 
-	return s.connectPeer(conn, inbound)
+	_ = s.outboundPeers.AddPeer(peer) // don't need to check error here
+	s.addresses.UpsertPeerAddr(peer)
+	return nil
 }
 
-func (s *server) connectPeer(conn net.Conn, inbound bool) error {
-	peer, err := peer.NewPeer(conn, inbound, s.config, s.chainParams, s.headersService, s.chainService, s.log)
+func (s *server) connectPeer(conn net.Conn, inbound bool) (*peer.Peer, error) {
+	peer := peer.NewPeer(conn, inbound, s.config, s.chainParams, s.headersService, s.chainService, s.log, s)
+
+	err := peer.Connect()
 	if err != nil {
-		return err
+		peer.Disconnect()
+		s.log.Error().Str("peer", peer.String()).
+			Bool("inbound", inbound).
+			Msgf("error connecting with peer, reason: %v", err)
+		return nil, err
 	}
 
-	s.peers = append(s.peers, peer)
+	peer.SendGetAddrInfo()
 
-	err = peer.Connect()
+	if !inbound {
+		err = peer.StartHeadersSync()
+		if err != nil {
+			peer.Disconnect()
+			return nil, err
+		}
+	}
```

Comment:
This part cannot be here, because it will start a sync with every peer we connect to. And we wanted to just sync from one peer.

I suggest, in the scope of this task, to start sync with the first connected peer, or with the first from the list or a random peer. And we have a [separate task SPV-744](https://bitcoinassociation.atlassian.net/browse/SPV-744) for choosing the best sync peer.

Reply:
What is wrong with syncing with multiple nodes?

Reply:
1. Unnecessary network load when syncing from 0. While syncing from 0, we ask for very old headers, which every node should have. So there's no need for asking multiple nodes to send us old headers, one is enough. Other nodes should be focused on keeping the network up to date with newest headers. Remember that when syncing we ask for batches of 2 000 headers every second or so. That's not a super small network load.
2. Unnecessary network load when synced. When synced and waiting for new headers, we could ask multiple nodes to send us the headers, so we can time the response and give a score to each peer based on their response time and then sync with just that one. But that's for another task IMHO.
3. DB load and synchronisation problems. When asking multiple peers for old nodes in batches, multiple peers are gonna send us 2 000 headers in more or less the same time every second. Then we have to put them into DB, and after the first batch is inserted all other batches are gonna return an error "headers already in db". And if we start blocking the db for synchronisation, we'll have race conditions and slow down the entire sync process significantly. Why would we do this to ourselves? 😬

Reply:
Changed. We use only one peer for sync from 0 now.